src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -1425,6 +1421,70 @@ func (s commitLogSource) maybeAddToIndex(
 	return err
 }
 
+func (s *commitLogSource) availability(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	var (
+		topoState                = runOpts.InitialTopologyState()
+		availableShardTimeRanges = result.ShardTimeRanges{}
+	)
+
+	for shardIDUint := range shardsTimeRanges {
+		shardID := topology.ShardID(shardIDUint)
+		hostShardStates, ok := topoState.ShardStates[shardID]
+		if !ok {
+			// This shard was not part of the topology when the bootstrapping
+			// process began.
+			continue
+		}
+
+		originHostShardState, ok := hostShardStates[topology.HostID(topoState.Origin.ID())]
+		if !ok {
+			// TODO(rartoul): Make this a hard error once we refactor the interface to support
+			// returning errors.
+			iOpts := s.opts.CommitLogOptions().InstrumentOptions()
+			invariantLogger := instrument.EmitInvariantViolationAndGetLogger(iOpts)
+			invariantLogger.Errorf(
+				"Initial topology state does not contain shard state for origin node and shard: %d", shardIDUint)
+			continue
+		}
+
+		originShardState := originHostShardState.ShardState
+		switch originShardState {
+		// In the Initializing and Unknown states we have to assume that the commit log
+		// is missing data and can't satisfy the bootstrap request.
+		case shard.Initializing:
+		case shard.Unknown:
+		// In the Leaving and Available case, we assume that the commit log contains
+		// all the data required to satisfy the bootstrap request because the node
+		// had (at some point) been completely bootstrapped for the requested shard.
+		// This doesn't mean that the node can't be missing any data or wasn't down
+		// for some period of time and missing writes in a multi-node deployment, it
+		// only means that technically the node has successfully taken ownership of
+		// the data for this shard and made it to a "bootstrapped" state which is
+		// all that is required to maintain our cluster-level consistency guarantees.
+		case shard.Leaving:
+			fallthrough
+		case shard.Available:
+			// Assume that we can bootstrap any requested time range, which is valid as
+			// long as the FS bootstrapper precedes the commit log bootstrapper.
+			// TODO(rartoul): Once we make changes to the bootstrapping interfaces
+			// to distinguish between "unfulfilled" data and "corrupt" data, then
+			// modify this to only say the commit log bootstrapper can fullfil
+			// "unfulfilled" data, but not corrupt data.
+			availableShardTimeRanges[shardIDUint] = shardsTimeRanges[shardIDUint]
+		default:
+			// TODO(rartoul): Make this a hard error once we refactor the interface to support
+			// returning errors.
+			panic(fmt.Sprintf("unknown shard state: %v", originShardState))
```

Comment:
Can we prioritize and make these `AvailableXyz(...)` methods return an error?  It would be really catastrophic if all the nodes panic when something does something stupid and modifies topology data in etcd in an unexpected way.

Reply:
Dont want to make the interface change in this P.R because it will double the size at the least, but I'll convert these to error logs and make the method return that it cant satisfy any ranges